High-level example of the JWT bearer flow with an operator token

jpengar opened the pull request:
#### What type of PR is this?

* documentation

#### What this PR does / why we need it:

This pull request adds a new section describing the JWT Bearer flow with Operator Token, including a sequence diagram and explanatory notes. This section outlines how the `sub` claim in the JWT assertion is set to the Operator Token and describes the interactions between the involved components.

#### Which issue(s) this PR fixes:

Fixes #326

#### Special notes for reviewers:

This pull request does not include the `openid` scope in the example sequence diagram (typo reported in issue #329 by @shilpa-padgaonkar)

#### Changelog input

```
 JWT bearer flow with Operator token
```

#### Additional documentation 

N/A

jpengar committed:
High-level example of the JWT bearer flow with an operator token

jpengar committed:
table of contents update

jpengar committed:
Address Diego's comments

jpengar commented on documentation/CAMARA-API-access-and-user-consent.md:
```diff
@@ -446,6 +447,50 @@ In case of Backend-to-backend scenarios where no user interaction is required (e
+end
+Note over BE,ExpO: Pre-requisite:<br> To verify the signed JWT assertion, the public key of the API Consumer must be shared<br> with API Provider as part of API Consumer's Application onboarding process
+Note over FE,BE: Feature needing Operator capability  
+Note over BE: API Consumer obtains a temporary token<br>following TS.43 definitions...
```
Fixed in https://github.com/camaraproject/IdentityAndConsentManagement/pull/334/commits/2c47b8ae20cf899a606633ce32c2a043f8941939

diegogonmar reviewed (approved):
LGTM

sebdewet reviewed (dismissed):
LGTM

jpengar committed:
Merge main branch

diegogonmar reviewed (approved):
LGTM

AxelNennker commented on documentation/CAMARA-API-access-and-user-consent.md:
```diff
@@ -391,6 +392,50 @@ The API Consumer uses the access token to invoke the CAMARA API, including it in
+      ES->>ES: Validate temporary token
+      ES->>ExpO: HTTP 200 OK <br>{operator_token}
+      ExpO->>ES: GetSubscriberDeviceInfo (ap2015, operator_token) 
+      ES->>ExpO: SubscriberDeviceInfo (MSISDN[,IMSI])
```
```suggestion
```

AxelNennker reviewed (changes requested):
These combined change requests show how this is done at DT.
The AZ creates an Eligibility Token that is used an an access token and validated at the ECS. The operatorToken then contains the MSISDN. The OperatorAccessToken then contains the MSISDN.

This reduces the flow by one network request compared to the example flow before my suggestions. We/DT do not call GetSubscriberInfo.

AxelNennker commented:
At DT the JWT Bearer Flow looks like this. This flow has one network request less the the sequence diagram proposal in this PR.
I suggest to have this example flow in 
##### JWT Bearer flow with Operator Token

```mermaid
sequenceDiagram
autonumber
title JWT Bearer Flow with Operator Token
participant FE as Device App<br>(Consumption Device)
participant BE as API Consumer<br>(Application Backend/Aggregator)
box API Provider / Operator
  participant ExpO as API Exposure Platform
  participant ES as Entitlement Server  
  participant Consent as Consent Master
end
Note over BE,ExpO: Pre-requisite:<br> To verify the signed JWT assertion, the public key of the API Consumer<br>must be shared with API Provider as part of<br>API Consumer's Application onboarding process
Note over FE,BE: Feature needing Operator capability  
Note over FE,BE: API Consumer obtains a temporary token<br>following TS.43 definitions...
alt Assertion Framework for OAuth 2.0 Client Authentication and Authorization Grants
    BE->>+ExpO: POST /token<br>grant_type=urn:ietf:params:oauth:grant-type:jwt-bearer<br>assertion=<br>{dpv:<purposeDpvValue> scope1 ... scopeN<br>sub=operatortoken:<temporaryToken>,...}<br>
    ExpO->>ExpO: Validate Authentication<br>request. Validate assertion, etc<br>
    ExpO->>ExpO: Create GSMA<br> ASAC.01 Eligibility Token<br>
    alt API Provider internal,<br>Inspired by TS.43 AcquireOperatorToken
      Note over ExpO,ES: Get information from Entitlement Server<br> needed for access_token creation<br>Scopes are adapted
      ExpO->>ES: AcquireOperatorToken(temporary_token=temporaryToken,<br>access_token=eligibilityToken)
      ES-->>ES: Validate eligibilityToken<br>
      ES->>ES: Validate temporary token<br>
      ES->>ES: Create operatorToken
      ES->>ExpO: HTTP 200 OK <br>{operatorToken}<br>
    end
    ExpO->>ExpO: Check legal basis of the purpose<br> e.g.: contract, legitimate_interest, consent, etc<br>
    opt If User Consent is required for the legal basis of the purpose  
        ExpO->>Consent: Check if Consent is granted
    end
    ExpO->>ExpO: Create 3-legged access token
    ExpO->>-BE: HTTP 200 OK<br>{"access_token": "{OperatorAccessToken}"}
end
BE->>ExpO: Send CAMARA API request<br>Authorization: Bearer {OperatorAccessToken}    
ExpO->>ExpO: Validate OperatorAccessToken,<br>grants Access,<br>gets API response  
ExpO->>BE: CAMARA API Response
Note over BE,FE: Response
```

DT's `subscriber tokens`, the response to AcquireTemporaryToken has this format:
Note: A subscriber token is a temporary token with routing information

```json
{
    "iss": "https://entitlement.csp.com/",
    "exp": 1639528912,
    "aud": "https://auth.csp.com",
    "iat": 1618354090,
    "jti": "dbe39bf3a3ba4238a513f51d6e1691c4",
    "kid": "s6BhdRkqt3",
    "auth": "https://auth.csp.com/",
    "api": "https://api.csp.com/",
    "operation_targets": ["GetPhoneNumber"],
    "temporary_token": "<JWE encrypted token>"
  }
```

DT's `Eligibility Token` looks like this, some parameters from the TS.43 AcquireOperatorToken request are inside the Eligibility Token and not parameters of this internal AcquireOperatorToken request with the same name.
The internal AcquireOperatorToken request has two parameters temporaryToken and accessToken.

```json
{
  "exp": 1762683439,
  "iat": 1762683379,
  "jti": "da9574da-ed2b-42d4-ae6c-4e58479a559a",
  "iss": "https://auth.csp.com/",
  "aud": "https://entitlement.csp.com",
  "scope": "MSISDN",
  "3rdParty_publicKey": {
    "kty": "EC",
    "crv": "P-256",
    "kid": "e1XGF-x-261ix5Fqgq2cuZWYv-aENPFai5lvvgWlL20",
    "x": "OzoyVCbj_UEeEj7Z7SB4mNjJQB682EH839dE480SPFQ",
    "y": "77LCGahreHLiyNgkPcgQPAoh-hwMGXFl4fQvcCo_Pow"
  },
  "client_id": "<client_id from JWT Bearer Flow assertion>"
}
```

As the AZ and the ECS are under the same authority, the CSP, the format of the Eligibility Token is something agreed upon and does not necessarily follow a standard. Especially, who the 3rdParty_publicKey belongs to or whether it exists is a choice of the Eligibility Service, here the Authorization Server.

All tokens are signed by their "iss".

@jpengar are you OK with using this sequence diagram? Then please copy the mermaid from this comment, maybe also the structure of the example temporary token and eligibility token for additional information.

jpengar committed:
Merge branch 'main' into jpengar/jwt-bearer-ot

jpengar committed:
Remove the specifics of TS.43

jpengar commented:
> @jpengar are you OK with using this sequence diagram? Then please copy the mermaid from this comment, maybe also the structure of the example temporary token and eligibility token for additional information.

@AxelNennker Following up on our discussion during the last ICM meeting, I have updated the PR to address your concerns regarding the internal interactions between the Auth Server and the Entitlement Server.

As agreed, we want to avoid prescribing implementation details that fall outside the scope of CAMARA. We also want to maintain consistency between CIBA and JWT Bearer flows with the operator token. To reflect this, I have (see https://github.com/camaraproject/IdentityAndConsentManagement/pull/334/commits/51cc4646c141a71a1110fa790abd2895882d8236):

- I have simplified the sequence diagram by removing the specific AcquireOperatorToken and GetSubscriberDeviceInfo calls. They have been replaced with a generic Note stating that these internal steps are beyond the scope of this document and may depend on each API Provider's implementation. 
- The descriptive text and final note have been slightly adjusted.

This approach maintains the necessary technical context while providing the flexibility for different operator architectures that you requested. Please let me know if you are satisfied with these changes.

shilpa-padgaonkar reviewed (approved):
LGTM
